lib/cache/PackFileCacheStrategy.js

```diff
@@ -112,170 +124,448 @@ class Pack {
+		}
+
+		// 5. GC and update location of merged items
+		const newLoc = this._findLocation();
+		this._gcAndUpdateLocation(mergedItems, mergedUsedItems, newLoc);
+
+		// 6. If not empty, store content somewhere
+		if (mergedItems.size > 0) {
+			this.content[newLoc] = new PackContent(
+				mergedItems,
+				mergedUsedItems,
+				memorize(async () => {
+					// unpack existing content in parallel
+					// after that values are accessible in .content
+					await Promise.all(mergedContent.map(c => c.unpack()));
+					/** @type {Map<string, any>} */
+					const map = new Map();
+					for (const content of mergedContent) {
+						for (const [identifier, value] of content.content) {
+							if (mergedItems.has(identifier)) {
+								map.set(identifier, value);
+							}
+						}
+					}
+					return map;
+				})
+			);
+			this.logger.log(
+				"Merged %d small files with %d cache items into pack %d",
+				mergedContent.length,
+				mergedItems.size,
+				newLoc
+			);
+		}
 	}
 
-	deserialize({ read, logger }) {
-		this.logger = logger;
-		this.etags = read();
-		this.unserializable = read();
-		// Mark the first 1% of the items for retry
-		// If they still fail they are added to the back again
-		// Eventually all items are retried, but not too many at the same time
-		if (this.unserializable.size > 0) {
-			let i = Math.ceil(this.unserializable.size / 100);
-			for (const item of this.unserializable) {
-				this.unserializable.delete(item);
-				if (--i <= 0) break;
+	/**
+	 * Split large content files with used and unused items
+	 * into two parts to separate used from unused items
+	 */
+	_optimizeUnusedContent() {
+		// 1. Find a large content file with used and unused items
+		for (let i = 0; i < this.content.length; i++) {
+			const content = this.content[i];
+			if (content === undefined) continue;
+			const size = content.getSize();
+			if (size < MIN_CONTENT_SIZE) continue;
+			const used = content.used.size;
+			const total = content.items.size;
+			if (used > 0 && used < total) {
+				// 2. Remove this content
+				this.content[i] = undefined;
+
+				// 3. Determine items for the used content file
+				const usedItems = new Set(content.used);
+				const newLoc = this._findLocation();
+				this._gcAndUpdateLocation(usedItems, usedItems, newLoc);
+
+				// 4. Create content file for used items
+				if (usedItems.size > 0) {
+					this.content[newLoc] = new PackContent(
+						usedItems,
+						new Set(usedItems),
+						async () => {
+							await content.unpack();
+							const map = new Map();
+							for (const identifier of usedItems) {
+								map.set(identifier, content.content.get(identifier));
+							}
+							return map;
+						}
+					);
+				}
+
+				// 5. Determine items for the unused content file
+				const unusedItems = new Set(content.items);
+				const usedOfUnusedItems = new Set();
+				for (const identifier of usedItems) {
+					unusedItems.delete(identifier);
+				}
+				const newUnusedLoc = this._findLocation();
```

Comment:
Just as idea, maybe store free locations indexes in separated structure with faster access (i.e. set of free indexes). Looks like `_findLocation` is frequent operation